rbi/stdlib/set.rbi

```diff
@@ -646,356 +646,6 @@ class SortedSet < Set
-#
-# Note that this operation leaves the array unchanged.
-#
-# ```ruby
-# arr = [1, 2, 3, 4, 5]
-# arr.each {|a| print a -= 10, " "}
-# # prints: -9 -8 -7 -6 -5
-# #=> [1, 2, 3, 4, 5]
-# ```
-#
-# Another sometimes useful iterator is
-# [`reverse_each`](https://docs.ruby-lang.org/en/2.7.0/Array.html#method-i-reverse_each)
-# which will iterate over the elements in the array in reverse order.
-#
-# ```ruby
-# words = %w[first second third fourth fifth sixth]
-# str = ""
-# words.reverse_each {|word| str += "#{word} "}
-# p str #=> "sixth fifth fourth third second first "
-# ```
-#
-# The [`map`](https://docs.ruby-lang.org/en/2.7.0/Array.html#method-i-map)
-# method can be used to create a new array based on the original array, but with
-# the values modified by the supplied block:
-#
-# ```ruby
-# arr.map {|a| 2*a}     #=> [2, 4, 6, 8, 10]
-# arr                   #=> [1, 2, 3, 4, 5]
-# arr.map! {|a| a**2}   #=> [1, 4, 9, 16, 25]
-# arr                   #=> [1, 4, 9, 16, 25]
-# ```
-#
-# ## Selecting Items from an [`Array`](https://docs.ruby-lang.org/en/2.7.0/Array.html)
-#
-# Elements can be selected from an array according to criteria defined in a
-# block. The selection can happen in a destructive or a non-destructive manner.
-# While the destructive operations will modify the array they were called on,
-# the non-destructive methods usually return a new array with the selected
-# elements, but leave the original array unchanged.
-#
-# ### Non-destructive Selection
-#
-# ```ruby
-# arr = [1, 2, 3, 4, 5, 6]
-# arr.select {|a| a > 3}       #=> [4, 5, 6]
-# arr.reject {|a| a < 3}       #=> [3, 4, 5, 6]
-# arr.drop_while {|a| a < 4}   #=> [4, 5, 6]
-# arr                          #=> [1, 2, 3, 4, 5, 6]
-# ```
-#
-# ### Destructive Selection
-#
-# [`select!`](https://docs.ruby-lang.org/en/2.7.0/Array.html#method-i-select-21)
-# and
-# [`reject!`](https://docs.ruby-lang.org/en/2.7.0/Array.html#method-i-reject-21)
-# are the corresponding destructive methods to
-# [`select`](https://docs.ruby-lang.org/en/2.7.0/Array.html#method-i-select) and
-# [`reject`](https://docs.ruby-lang.org/en/2.7.0/Array.html#method-i-reject)
-#
-# Similar to
-# [`select`](https://docs.ruby-lang.org/en/2.7.0/Array.html#method-i-select) vs.
-# [`reject`](https://docs.ruby-lang.org/en/2.7.0/Array.html#method-i-reject),
-# [`delete_if`](https://docs.ruby-lang.org/en/2.7.0/Array.html#method-i-delete_if)
-# and
-# [`keep_if`](https://docs.ruby-lang.org/en/2.7.0/Array.html#method-i-keep_if)
-# have the exact opposite result when supplied with the same block:
-#
-# ```ruby
-# arr.delete_if {|a| a < 4}   #=> [4, 5, 6]
-# arr                         #=> [4, 5, 6]
-#
-# arr = [1, 2, 3, 4, 5, 6]
-# arr.keep_if {|a| a < 4}   #=> [1, 2, 3]
-# arr                       #=> [1, 2, 3]
-# ```
-#
-# for pack.c
-class Array
-  sig { params(blk: T.nilable(T.proc.params(arg0: T.untyped).void)).returns(T::Set[T.untyped]) }
-  def to_set(&blk); end
-end
-
```

Comment:
I think this will work fine, but we should also add a `sig` on the `Enumerable#to_set` to match the one that is being removed from `Array#to_set`. I guess the tests would be expecting the specific `Set` return.

Reply:
At the same time, `Enumerable#to_set` is a really weird method, if you use the `klass` argument and such:
```ruby
[5] pry(main)> class Foo; def initialize(x); @x = x end; end
[6] pry(main)> [1,2,3].to_set(Foo)
=> #<Foo:0x00007fd88d8f1598 @x=[1, 2, 3]>
```

Can I suggest that we forbid the usage of `klass` and `args` in `Enumerable#to_set` by intentionally making the sig `def to_set(&blk); end`?

Reply:
Oh, I see what the intent of the `klass` parameter is now. It is there to support `to_set(SortedSet)` and similar `Set` subclass usages. But, since Ruby does not have a way to codify that type requirement 😉, it is not explicit in the code. So, I suggest the following signature for `Enumerable#to_set`:
```ruby
module Enumerable
  # Makes a set from the enumerable object with given arguments. Needs to
  # +require "set"+ to use this method.
  sig { 
    type_parameters(:U, :V)
    .params(
      klass: T::Set[T.type_parameter(:U)],
      args: T.untyped,
      block: T.nilable(T.proc.params(arg0: T.type_parameter(:U)).returns(T.type_parameter(:V)))
    )
    .returns(T::Set[T.type_parameter(:V)]) 
  }
  def to_set(klass = _, *args, &block); end
end

Reply:
With your edit I get

```
Malformed type declaration. Generic class without type arguments `Set` (fix available)
```

Do you have a suggestion here?

Reply:
(my default action right now is to leave it without a sig)

Reply:
Yeah, I think I really should have said `T.class_of(Set)`. I am fine either way with or without sigs for the moment. I can test and add a sig later if you want.